x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/OffsetSourceFieldMapper.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.mapper;
+
+import org.apache.lucene.index.FieldInfos;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.index.fielddata.FieldDataContext;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.mapper.DocumentParserContext;
+import org.elasticsearch.index.mapper.FieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.MapperBuilderContext;
+import org.elasticsearch.index.mapper.TextSearchInfo;
+import org.elasticsearch.index.mapper.ValueFetcher;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.search.fetch.StoredFieldsSpec;
+import org.elasticsearch.search.lookup.Source;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+
+/**
+ * A {@link FieldMapper} that maps a field name to its start and end offsets.
+ * The {@link CharsetFormat} used to compute the offsets is specified via the charset parameter.
+ * Currently, only {@link CharsetFormat#UTF_16} is supported, aligning with Java's {@code String} charset
+ * for simpler internal usage and integration.
+ *
+ * Each document can store at most one value in this field.
+ *
+ * Note: This mapper is not yet documented and is intended exclusively for internal use by
+ * {@link SemanticTextFieldMapper}. If exposing this mapper directly to users becomes necessary,
+ * extending charset compatibility should be considered, as the current default (and sole supported charset)
+ * was chosen for ease of Java integration.
+ */
+public class OffsetSourceFieldMapper extends FieldMapper {
```

Comment:
should we implement doValidate() to ensure that the `field` part refers to an actual field, and the field type is a text field?

Reply:
After giving it more thought, I don't think this is necessary at this point. There's nothing preventing us from using this offset source to reference a non-mapped field. Since it's intended for internal use only, I'm not particularly concerned about it.

Reply:
Thanks for checking this Jim!